example/reduce/src/iterator.hpp

```diff
@@ -0,0 +1,422 @@
+/**
+ * \file
+ * Copyright 2018 Jonas Schenke
+ *
+ * This file is part of alpaka.
+ *
+ * alpaka is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * alpaka is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with alpaka.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include <alpaka/alpaka.hpp>
+
+//#############################################################################
+//! An iterator base class.
+//!
+//! \tparam T The type.
+//! \tparam TBuf The buffer type (standard is T).
+template <typename T, typename TBuf = T>
+class Iterator
+{
+protected:
+    const TBuf *data;
+    uint64_t index;
+    const uint64_t maximum;
+
+public:
+    //-----------------------------------------------------------------------------
+    //! Constructor.
+    //!
+    //! \param data A pointer to the data.
+    //! \param linearizedIndex The linearized index.
+    //! \param gridSize The grid size.
+    //! \param n The problem size.
```

Comment:
The comment does not fully match the function parameters (probably was not updated)